### PR TITLE
[IMP] sale_project: compute the amount invoiced from the down payments for the project profitability

### PR DESCRIPTION
**Pull request by damrOdoo**
[IMP] sale_project: compute the amount invoiced from the down payments for the project profitability

This commit's purpose is to take into account the down payments for the
project profitability of the project update window.
Currently, if an SO has a total of 300$, and a down payment is created for 100$, the project
profitability will not compute the 100$ into the 'amount_invoiced'
column. After this commit, it will. A negative value of -100 is also
added in the 'amount_to_invoice' column in order to keep the total of
that column, and the total of the 'expected' column of the project
update consistent with the total of the SO.

task-3204827

### CODE REVIEW

File: addons/sale_project/models/project.py [hunk 1]
```diff
@@ -478,15 +505,8 @@ def _get_revenues_items_from_invoices(self, excluded_move_line_ids=None):
...
-        revenue_items_from_sol = self._get_revenues_items_from_sol(
-            domain,
-            with_action,
-        )
         revenues = profitability_items['revenues']
-        revenues['data'] += revenue_items_from_sol['data']
-        revenues['total']['to_invoice'] += revenue_items_from_sol['total']['to_invoice']
-        revenues['total']['invoiced'] += revenue_items_from_sol['total']['invoiced']
-
+        self._get_revenues_items_from_sol(revenues, domain, with_action)
```
Comment (auon-odoo) [inline addons/sale_project/models/project.py:509]: Was it necessary to change that? :thinking:

**Review by auon-odoo — CHANGES_REQUESTED** [review]
Thank you! :fire:

File: addons/sale_project/models/project.py [hunk 2]
```diff
@@ -330,30 +332,54 @@ def _get_profitability_sale_order_items_domain(self, domain=None):
...
+                    sols_per_product[res['product_id'][0]] = (
+                            res['untaxed_amount_to_invoice'],
+                            res['untaxed_amount_invoiced'],
+                            res['ids'],
+                        )
+            if down_payment_amount_invoiced:
+                dp_data = {'id': 'downpayments', 'sequence': sequence_per_invoice_type['downpayments'], 'invoiced': down_payment_amount_invoiced, 'to_invoice': -down_payment_amount_invoiced}
+                if with_action and self.user_has_groups('sales_team.group_sale_salesman_all_leads'):
+                    sale_order_items = self.env['sale.order.line'].browse(down_payment_sol_ids)
+                    if sale_order_items:
```
[the remark this answers is not in the document]
Comment (damrOdoo) [reply addons/sale_project/models/project.py:369]: No you did not miss anything :D. The browse was indeed not needed. The same for the following condition 'if sale_order_item'.

File: addons/sale_project/models/project.py [hunk 3]
```diff
@@ -340,20 +341,49 @@ def _get_profitability_sale_order_items_domain(self, domain=None):
...
+            for res in sale_line_read_group:
+                if res['is_downpayment']:
+                    downpayment_amount_invoiced += res['untaxed_amount_invoiced']
+                    downpayment_sol_ids += res['ids']
+                else:
+                    sols_per_product[res['product_id'][0]] = (
+                            res['untaxed_amount_to_invoice'],
+                            res['untaxed_amount_invoiced'],
+                            res['ids'],
+                        )
```
Comment (auon-odoo) [inline addons/sale_project/models/project.py:365]:
```suggestion
                    sols_per_product[res['product_id'][0]] = (
                            res['untaxed_amount_to_invoice'],
                        res['untaxed_amount_invoiced'],
                        res['ids'],
                    )
```
:angel:

File: addons/sale_project/models/project.py [hunk 4]
```diff
@@ -340,20 +341,49 @@ def _get_profitability_sale_order_items_domain(self, domain=None):
...
+            if downpayment_amount_invoiced:
+                downpayments_data = {
+                    'id': 'downpayments',
+                    'sequence': sequence_per_invoice_type['downpayments'],
+                    'invoiced': downpayment_amount_invoiced,
+                    'to_invoice': -downpayment_amount_invoiced
+                }
+                if with_action and self.user_has_groups('sales_team.group_sale_salesman_all_leads'):
```
Comment (auon-odoo) [inline addons/sale_project/models/project.py:373]: Why don't we need to check if there are some ids? :thinking:

File: addons/sale_project/models/project.py [hunk 5]
```diff
@@ -387,19 +417,18 @@ def _get_revenues_items_from_sol(self, domain=None, with_action=True):
...
+                    args = [section_name, [('id', 'in', sale_order_items.ids)]]
+                    if len(sale_order_items) == 1:
+                        args.append(sale_order_items.id)
+                    action_params = {
+                        'name': 'action_profitability_items',
+                        'type': 'object',
+                        'args': json.dumps(args),
+                    }
+                    other_revenues['action'] = action_params
+        data += [{'id': invoice_type, 'sequence': sequence_per_invoice_type[invoice_type], **vals} for invoice_type, vals in revenues_dict.items()]
```
Comment (auon-odoo) [inline addons/sale_project/models/project.py:429]:
```suggestion
        data += [{
            'id': invoice_type,
            'sequence': sequence_per_invoice_type[invoice_type],
            **vals
         } for invoice_type, vals in revenues_dict.items()]
```
Tant qu'on y est :angel:

File: addons/sale_project/tests/test_project_profitability.py [hunk 6]
```diff
@@ -234,13 +240,80 @@ def test_project_profitability(self):
...
+        self._assert_dict_equal(invoice_type, sequence_per_invoice_type, material_order_line, service_sols, manual_service_order_line, down_payment_invoiced)
+
+        # Create a second down payment for a fixed amount of 115
+        downpayment = self.env['sale.advance.payment.inv'].with_context(context).create({
+            'advance_payment_method': 'fixed',
+            'fixed_amount': 115,
+            'deposit_account_id': self.company_data['default_account_revenue'].id
+        })
+        # the down payment is with the amount with the default 15% included, therefore, the amount of the dp is higher than the amount of the sol created.
+        down_payment_invoiced = 2*down_payment_invoiced
```
Comment (auon-odoo) [inline addons/sale_project/tests/test_project_profitability.py:269]:
```suggestion
        down_payment_invoiced = 2 * down_payment_invoiced
```

**Review by auon-odoo — CHANGES_REQUESTED** [review]
Some style, and a question. And then, we'll be good to go! Thank you :pray:

File: addons/sale_project/models/project.py (hunk 4 above)
Comment (damrOdoo) [reply addons/sale_project/models/project.py:373]: There's already an if statement above to ensure there is an amount for the 'downpayments' section. I can change the 'if downpayment_amount_invoiced' to 'if downpayment_sol_ids' if you think its neater that way.

Comment (auon-odoo) [reply addons/sale_project/models/project.py:373]: Ho ok I see. :astonished:
Hm, maybe, yes. Because, in case (now or in the future) we can have downpayments of 0, checking the id rather than the amount sounds more solid. What do you think?

Comment (damrOdoo) [reply addons/sale_project/models/project.py:373]: well, for now at least, it is not possible. Because the downpayments SOL are locked and can not be modified. And if it get changed, i think its better to do the check with the amount, because i dont see the utility of adding a section in the project updates, with an amount to invoice of 0. Might as well just not show the section. There are already too many sections in that panel (imo)

Comment (auon-odoo) [reply addons/sale_project/models/project.py:373]: You're right :ok_hand:

File: addons/sale_project/tests/test_project_profitability.py [hunk 7]
```diff
@@ -234,13 +240,80 @@ def test_project_profitability(self):
                 },
             },
         )
+        # Create a down payment for a fixed amount of 115
+        context = {
+            'active_model': 'sale.order',
+            'active_ids': [self.sale_order.id],
```
Comment (auon-odoo) [inline addons/sale_project/tests/test_project_profitability.py:246]:
```suggestion
            'active_ids': self.sale_order.ids,
```
:angel:

File: addons/sale_project/models/project.py [hunk 8]
```diff
@@ -486,7 +519,6 @@ def _get_profitability_items(self, with_action=True):
         revenues['data'] += revenue_items_from_sol['data']
         revenues['total']['to_invoice'] += revenue_items_from_sol['total']['to_invoice']
         revenues['total']['invoiced'] += revenue_items_from_sol['total']['invoiced']
-
```
Comment (auon-odoo) [inline addons/sale_project/models/project.py:489]: This is a useless diff :angel:

File: addons/sale_project/tests/test_project_profitability.py [hunk 9]
```diff
@@ -234,13 +240,80 @@ def test_project_profitability(self):
...
+            'active_id': self.sale_order.id,
+            'default_journal_id': self.company_data['default_journal_sale'].id,
+        }
```
Comment (auon-odoo) [inline addons/sale_project/tests/test_project_profitability.py:249]:
```suggestion
        Downpayment = self.env['sale.advance.payment.inv'].with_context({
            'active_model': 'sale.order',
            'active_ids': self.sale_order.ids,
            'active_id': self.sale_order.id,
            'default_journal_id': self.company_data['default_journal_sale'].id,
        })
```
This is also possible, and slightly preferable imho :slightly_smiling_face:

File: addons/sale_project/tests/test_project_profitability.py [hunk 10]
```diff
@@ -234,13 +240,80 @@ def test_project_profitability(self):
...
+        downpayment = self.env['sale.advance.payment.inv'].with_context(context).create({
+            'advance_payment_method': 'fixed',
+            'fixed_amount': 115,
+            'deposit_account_id': self.company_data['default_account_revenue'].id
+        })
+        # the down payment is with the amount with the default 15% included, therefore, the amount of the dp is higher than the amount of the sol created.
```
Comment (auon-odoo) [inline addons/sale_project/tests/test_project_profitability.py:255]: While we are at it, please standardize your comment, capital-letter and punctuation wise :pray:

**Review by auon-odoo — CHANGES_REQUESTED** [review]
Thanks again :ok_hand:

File: addons/sale_project/models/project.py [hunk 11]
```diff
@@ -345,20 +346,49 @@ def _get_profitability_sale_order_items_domain(self, domain=None):
...
+                if with_action and self.user_has_groups('sales_team.group_sale_salesman_all_leads'):
+                    args = ['downpayments', [('id', 'in', downpayment_sol_ids)]]
+                    if len(downpayment_sol_ids) == 1:
+                        args.append(downpayment_sol_ids)
+                    action_params = {
+                        'name': 'action_profitability_items',
+                        'type': 'object',
+                        'args': json.dumps(args),
+                    }
+                    downpayments_data.update({'action': action_params})
```
Comment (auon-odoo) [inline addons/sale_project/models/project.py:387]:
```suggestion
                    downpayments_data['action'] = {
                        'name': 'action_profitability_items',
                        'type': 'object',
                        'args': json.dumps(args),
                    }
```

File: addons/sale_project/tests/test_project_profitability.py [hunk 12]
```diff
@@ -228,19 +234,86 @@ def test_project_profitability(self):
...
+                            'invoiced': material_order_line.untaxed_amount_invoiced,
+                        },
+                    ],
+                    'total': {
+                        'to_invoice': sum(service_sols.mapped(
+                            'untaxed_amount_to_invoice')) + material_order_line.untaxed_amount_to_invoice - down_payment_invoiced,
+                        'invoiced': manual_service_order_line.untaxed_amount_invoiced + material_order_line.untaxed_amount_invoiced + down_payment_invoiced,
+                    },
+                },
+                'costs': {  # no cost because we have no purchase orders.
```
Comment (auon-odoo) [inline addons/sale_project/tests/test_project_profitability.py:310]:
```suggestion
                'costs': {  # No cost because we have no purchase orders.
```

**Review by auon-odoo — APPROVED** [review]
Lgtm :zap: :green_circle: :bird:

File: addons/sale_project/models/project.py [hunk 13]
```diff
@@ -370,20 +371,49 @@ def _get_profitability_sale_order_items_domain(self, domain=None):
...
+            for res in sale_line_read_group:
+                if res[1]:
+                    downpayment_amount_invoiced += res[4]
+                    downpayment_sol_ids += res[2]
+                else:
+                    sols_per_product[res[0].id] = (
+                        res[3],
+                        res[4],
+                        res[2],
+                    )
```
Comment (xavierbol) [inline addons/sale_project/models/project.py:395]: May I ask you to "destructure" the tuple to name each element in that tuple. 😇 

```suggestion
            for product_id, is_downpayment, sol_ids, total_amount_to_invoice, total_amount_invoiced in sale_line_read_group:
                if is_downpayment:
                    downpayment_amount_invoiced += res[4]
                    downpayment_sol_ids += res[2]
                else:
                    sols_per_product[product_id.id] = (
                        total_amount_to_invoice,
                        total_amount_invoiced,
                        sol_ids,
                    )
```

It will inscrease the readability of your code. 🙂

File: addons/sale_project/models/project.py (hunk 11 above)
Comment (xavierbol) [reply addons/sale_project/models/project.py:413]: We can indeed do Audric's suggestion. 🙂

File: addons/sale_project/models/project.py [hunk 14]
```diff
@@ -427,8 +457,13 @@ def _get_revenues_items_from_sol(self, domain=None, with_action=True):
                         action_params['res_id'] = sale_order_items.id
                     materials['action'] = action_params
         sequence_per_invoice_type = self._get_profitability_sequence_per_invoice_type()
+        data += [{
+            'id': invoice_type,
+            'sequence': sequence_per_invoice_type[invoice_type],
+            **vals
```
Comment (xavierbol) [inline addons/sale_project/models/project.py:463]: Nitpicking. 😇 

```suggestion
            **vals,
```

File: addons/sale_project/models/project.py [hunk 15]
```diff
@@ -517,7 +552,7 @@ def _add_invoice_items(self, domain, profitability_items, with_action=True):
         )[0][0]
         revenue_items_from_invoices = self._get_revenues_items_from_invoices(
             excluded_move_line_ids=sale_lines.invoice_lines.ids,
-            with_action=with_action
+            with_action=with_action,
```
Comment (xavierbol) [inline addons/sale_project/models/project.py:555]: Unnecessary diff (it is to avoid to change the last commit in that line just for that change)

File: addons/sale_project/tests/test_project_profitability.py [hunk 16]
```diff
@@ -42,6 +42,12 @@ def setUpClass(cls):
             'service_tracking': 'task_global_project',
             'project_id': cls.project.id,
         })
+        cls.down_payment_product = cls.env['product.product'].create({
+            'name': "downpayment product, used to simulate down payments",
+            'standard_price': 30,
+            'type': 'service',
+            'service_policy': 'ordered_prepaid'
```
Comment (xavierbol) [inline addons/sale_project/tests/test_project_profitability.py:49]: Nitpicking. 😇 
```suggestion
            'service_policy': 'ordered_prepaid',
```

File: addons/sale_project/tests/test_project_profitability.py [hunk 17]
```diff
@@ -308,6 +314,37 @@ def test_project_profitability(self):
...
+        Downpayment = {
+            'active_model': 'sale.order',
+            'active_ids': self.sale_order.ids,
+            'active_id': self.sale_order.id,
+            'default_journal_id': self.company_data['default_journal_sale'].id,
+        }
+        downpayment = self.env['sale.advance.payment.inv'].with_context(Downpayment).create({
+            'advance_payment_method': 'fixed',
+            'fixed_amount': 115,
+            'deposit_account_id': self.company_data['default_account_revenue'].id
```
Comment (xavierbol) [inline addons/sale_project/tests/test_project_profitability.py:327]: Nitpicking. 

```suggestion
            'deposit_account_id': self.company_data['default_account_revenue'].id,
```

File: addons/sale_project/tests/test_project_profitability.py [hunk 18]
```diff
@@ -308,6 +314,37 @@ def test_project_profitability(self):
...
+        downpayment.create_invoices()
+        self.sale_order.invoice_ids[2].action_post()
+        # Ensures the down payment is correctly computed for the project profitability.
+        self._assert_dict_equal(invoice_type, sequence_per_invoice_type, material_order_line, service_sols, manual_service_order_line, down_payment_invoiced)
+
+        # Create a second down payment for a fixed amount of 115.
+        downpayment = self.env['sale.advance.payment.inv'].with_context(Downpayment).create({
+            'advance_payment_method': 'fixed',
+            'fixed_amount': 115,
+            'deposit_account_id': self.company_data['default_account_revenue'].id
```
Comment (xavierbol) [inline addons/sale_project/tests/test_project_profitability.py:341]: Nitpicking. 😇 

```suggestion
            'deposit_account_id': self.company_data['default_account_revenue'].id,
```

File: addons/sale_project/tests/test_project_profitability.py [hunk 19]
```diff
@@ -326,6 +363,48 @@ def test_project_profitability(self):
...
+                    'data': [
+                        {
+                            'id': 'other_revenues',
+                            'sequence': sequence_per_invoice_type['other_revenues'],
+                            'invoiced': 100.0,
+                            'to_invoice': 0.0,
+                        },
+                        {
+                            'id': 'downpayments', 'sequence': 20, 'invoiced': down_payment_invoiced,
+                            'to_invoice': -down_payment_invoiced
```
Comment (xavierbol) [inline addons/sale_project/tests/test_project_profitability.py:380]: Nitpicking. 😇 

```suggestion
                            'to_invoice': -down_payment_invoiced,
```

**Review by xavierbol — COMMENTED** [review]
Thanks for your work, I left some comments. 🙂

File: addons/sale_project/models/project.py (hunk 15 above)
Comment (xavierbol) [reply addons/sale_project/models/project.py]: IMO, we usually blame the code in the business and not in the test.

But indeed, if we want to do it everywhere, we should apply it in the 2 lines you pointed in the test. 🙂 

It is up to you, it is not really mandatory. 🙂

**Review by xavierbol — APPROVED** [review]
LGTM, thanks for your work. 🙂 

robodoo r+
